src/main/java/com/conveyal/datatools/manager/jobs/MergeLineContext.java

```diff
@@ -0,0 +1,899 @@
+            if (hasDuplicateError(idErrors)) skipRecord = true;
+        }
+
+        if (newAgencyId != null && field.name.equals(AGENCY_ID)) {
+            LOG.info(
+                "Updating route#agency_id to (auto-generated) {} for route={}",
+                newAgencyId, keyValue);
+            val = newAgencyId;
+        }
+    }
+
+    private boolean hasPrimaryKeyErrors(Set<NewGTFSError> primaryKeyErrors) {
+        return "".equals(keyValue) && field.name.equals(table.getKeyFieldName()) && hasDuplicateError(primaryKeyErrors);
+    }
+
+    private boolean useAltKey() {
+        return keyField.equals("stop_code") || keyField.equals("route_short_name");
+    }
+
+    public void checkMissingAgencyId() {
+        if (table.name.equals(Table.AGENCY.name) && (keyFieldMissing || keyValue.equals(""))) {
+            // agency_id is optional if only one agency is present, but that will
+            // cause issues for the feed merge, so we need to insert an agency_id
+            // for the single entry.
+            newAgencyId = UUID.randomUUID().toString();
+            if (keyFieldMissing) {
+                // Only add agency_id field if it is missing in table.
+                List<Field> fieldsList = new ArrayList<>(Arrays.asList(fieldsFoundInZip));
+                fieldsList.add(Table.AGENCY.fields[0]);
+                fieldsFoundInZip = fieldsList.toArray(fieldsFoundInZip);
+                allFields.add(Table.AGENCY.fields[0]);
+            }
+            fieldsFoundList = Arrays.asList(fieldsFoundInZip);
+        }
+    }
+
+    public void checkStopCodeStuff() throws IOException {
+        if (shouldCheckStopCodes()) {
+            // Before reading any lines in stops.txt, first determine whether all records contain
+            // properly filled stop_codes. The rules governing this logic are as follows:
+            // 1. Stops with location_type greater than 0 (i.e., anything but 0 or empty) are permitted
+            //    to have empty stop_codes (even if there are other stops in the feed that have
+            //    stop_code values). This is because these location_types represent special entries
+            //    that are either stations, entrances/exits, or generic nodes (e.g., for
+            //    pathways.txt).
+            // 2. For regular stops (location_type = 0 or empty), all or none of the stops must
+            //    contain stop_codes. Otherwise, the merge feeds job will be failed.
+            int stopsMissingStopCodeCount = 0;
+            int stopsCount = 0;
+            int specialStopsCount = 0;
+            int locationTypeIndex = getFieldIndex(fieldsFoundInZip, "location_type");
+            int stopCodeIndex = getFieldIndex(fieldsFoundInZip, "stop_code");
+            // Get special stops reader to iterate over every stop and determine if stop_code values
+            // are present.
+            CsvReader stopsReader = table.getCsvReader(feed.zipFile, null);
```

Comment:
Please close this reader.

Reply:
Handled in https://github.com/ibi-group/datatools-server/pull/376/commits/e90b8c1e126267f0984a139725010523284e9136.